### PR TITLE
Support primitive type of state for store

### PR DESCRIPTION
**Pull request by andy-yuanchang**

### CODE REVIEW

File: packages/redox/__tests__/defineModel.test.ts [hunk 1]
```diff
@@ -52,15 +52,70 @@ describe('defineModel worked:', () => {
...
 				})
-			}).toThrow()
+			}).not.toThrow()
+		})
+
+		test('state could be a string', () => {
+			expect(() => {
+				const model = defineModel({
+					name: 'a',
+					// @ts-ignore
```
Comment (wangjinyang) [inline packages/redox/__tests__/defineModel.test.ts]: // @ts-ignore 可以去掉了啊

File: packages/redox/__tests__/redoxStore.test.ts [hunk 2]
```diff
@@ -32,6 +32,71 @@ describe('redox worked:', () => {
 		expect(store.$state()).toEqual({ value: 2 })
 	})
 
+	test('initialize state as a String should work', () => {
```
Comment (wangjinyang) [inline packages/redox/__tests__/redoxStore.test.ts]: initialize state as a String should work 和 initialize state as a Number should work 写反了吧

File: packages/redox/src/views/index.ts [hunk 3]
```diff
@@ -84,27 +87,46 @@ const getStateCollection = () => {
...
-							viewNode.set(args, res)
+					if (p === '$state') {
+						const $state = result
+						result = function () {
+							let state = $state()
+							compareTree.set($state, {
+								children: state,
+							})
+
+							isCollectionKeys = true
```
Comment (wangjinyang) [inline packages/redox/src/views/index.ts]: isCollectionKeys = true 不需要吧，没有打断收集

File: packages/redox/src/views/index.ts [hunk 4]
```diff
@@ -122,7 +144,29 @@ const getStateCollection = () => {
...
 function compareObject(prevObj: any, nextObj: any, compare: ICompare) {
 	if (!isComplexObject(prevObj)) {
-		// sample value like string number just call ===
+		// $state function comparison
+		if (typeof prevObj === 'function') {
+			const treeMap = compare.tree.get(prevObj)
+			if (treeMap) {
+				const child = treeMap!.children
+				const nextChild = nextObj()
+				const comparedChild = compare.tree.get(child)?.children
```
Comment (wangjinyang) [inline packages/redox/src/views/index.ts]: 这是一个递归函数，是不是直接return compareObject(childObj, nextChildObj, compare) 就可以了

File: packages/redox/src/validate.ts [hunk 5]
```diff
@@ -42,14 +42,9 @@ const validate = (runValidations: () => Validation[]): void => {
...
 export const validateModel = (model: AnyModel): void => {
-	validate(() => [
-		[!model, 'model config is required'],
-		[
-			typeof model.state !== 'object',
-			'model "state" is required and it should be object !',
-		],
-	])
-	const keys = new Set<string>(Object.keys(model.state))
+	validate(() => [[!model, 'model config is required']])
```
Comment (wangjinyang) [inline packages/redox/src/validate.ts:45]: 不支持symbol和biginit这里缺少校验

File: packages/redox/__tests__/actions.test.ts [hunk 6]
```diff
@@ -274,6 +291,77 @@ describe('actions worked:', () => {
...
+		const addedValue: number = 2
+		const modifier = (state: any) => {
+			state += addedValue
+			return state + addedValue
+		}
+		store.makeCall(modifier)
+		expect(originalValue).toEqual(store.$state())
+	})
+
+	test('this.$modify should modify nothing if typeof state is string', () => {
```
Comment (liximomo) [inline packages/redox/__tests__/actions.test.ts]: 这个测试合 'this.$modify should modify nothing if typeof state is number', 重复了，他们算是一个case的不同实例，但本质上应该算是重复的

File: packages/redox/__tests__/actions.test.ts [hunk 7]
```diff
@@ -673,6 +790,95 @@ describe('actions worked:', () => {
 			store.makeCall()
 		})
 
+		test("$dep's $modify should modify nothing if typeof state is number", () => {
```
Comment (liximomo) [inline packages/redox/__tests__/actions.test.ts]: 感觉这个也和前面的this.$modify重复了？

**Review by liximomo — CHANGES_REQUESTED** [review]

File: packages/redox/__tests__/actions.test.ts (hunk 6 above)
Comment (andy-yuanchang) [reply packages/redox/__tests__/actions.test.ts]: 那this.$modify should modify nothing if typeof state is boolean也需要合併成一個case嗎？

File: packages/redox/src/views/index.ts (hunk 4 above)
Comment (andy-yuanchang) [reply packages/redox/src/views/index.ts]: 確實可以！而且state為array時產生的cache問題也解決了，不過確切原因還不知道，有待釐清，謝謝～

File: packages/redox/__tests__/redoxStore.test.ts (hunk 2 above)
Comment (andy-yuanchang) [reply packages/redox/__tests__/redoxStore.test.ts]: 應該沒有啊
as a String用'initial'做初始
as a Number用1做初始

File: packages/redox/__tests__/defineModel.test.ts (hunk 1 above)
Comment (andy-yuanchang) [reply packages/redox/__tests__/defineModel.test.ts]: 已修正

File: packages/redox/__tests__/redoxStore.test.ts (hunk 2 above)
Comment (wangjinyang) [reply packages/redox/__tests__/redoxStore.test.ts]: 这里你再看下和你讲的是相反的吧


	test('initialize state as a String should work', () => {
		manager = redox({
			initialState: {
				countNumber: 1,
			},
		})
		const countNumber = defineModel({
			name: 'countNumber',
			state: 0,
			reducers: {
				increment(state) {
					return ++state
				},
			},
		})

		const store = manager.get(countNumber)
		expect(store.$state()).toEqual(1)
		store.increment()
		expect(store.$state()).toEqual(2)
	})

	test('initialize state as a Number should work', () => {
		manager = redox({
			initialState: {
				textModel: 'initial',
			},
		})
		const textModel = defineModel({
			name: 'textModel',
			state: '',
			reducers: {
				append(state, payload) {
					return `${state}${payload}`
				},
				replace(_, payload) {
					return payload
				},
			},
		})

		const store = manager.get(textModel)
		expect(store.$state()).toEqual('initial')
		store.append('_test')
		expect(store.$state()).toEqual('initial_test')
	})

Comment (andy-yuanchang) [reply packages/redox/__tests__/redoxStore.test.ts]: 我最新的commit應該有包含這修改才對

File: packages/redox/__tests__/actions.test.ts (hunk 6 above)
Comment (liximomo) [reply packages/redox/__tests__/actions.test.ts]: 对的

**Review by liximomo — APPROVED** [review]
/lgtm

**Pull request merged**